File: Microsoft/754.reach-a-number.go
package leetcode

// ReachNumber754 uses the fact that after k moves the farthest right position
// is S=1+...+k. Flipping selected moves reduces S by twice their sum, so target
// is reachable iff S>=target and S-target is even.
//
// Time: O(sqrt(target)). Space: O(1).
func ReachNumber754(target int) int {
	target = absInt(target)
	step, total := 0, 0
	for total < target || (total-target)%2 != 0 {
		step++
		total += step
	}
	return step
}

File: companies/Ozon/754.reach-a-number.java
/**
 * Algorithm:
 * Move in the positive direction until the triangular sum reaches target and
 * the difference has even parity. Flipping the sign of one step changes the
 * final position by twice that step, so only parity matters after overshooting.
 *
 * Complexity:
 * Time O(sqrt(target)), space O(1).
 */
class Solution {
    public int reachNumber(int target) {
        target = Math.abs(target);
        int step = 0;
        int total = 0;
        while (total < target || (total - target) % 2 != 0) {
            step++;
            total += step;
        }
        return step;
    }
}

